### PR TITLE
test: make heavy attention test files more efficient (bulk JIT precompile + faster references)

### PR DESCRIPTION
**Pull request by bkryu**
## 📌 Description

The three slowest files in the H100 unit-test CI job (`task_run_unit_tests.sh`) spend 85–95% of their wall time on serial first-use JIT compilation, and most of the rest in slow host-side reference code — not in the kernels under test. This PR makes them more efficient without changing what is tested: same kernels, same parametrizations, same tolerances

**1. Bulk JIT precompilation (`tests/attention/conftest.py`)**

Generalizes the existing XQA prebuild hook into a per-file registry covering `test_hopper_fp8_attention.py`, `test_fmha_v2_prefill.py`, and `test_batch_prefill_kernels.py` (XQA behavior unchanged). 

For each registered file the hook derives the unique set of JitSpecs its collected tests will need, compiles them as one parallel ninja graph via `build_jit_specs`, and hardlinks the built `.so`s into the AOT dir so test-time loads skip the per-module ninja dependency scan. 

Collectors mirror each test's own arch skip guards (SM90a-only for hopper fp8; SM90a/SM12x split for fmha_v2; SM80+/SM100+ gates for large head dims and fp8/NVFP4 KV), so non-applicable hardware builds nothing. **Any failure falls back to normal per-test JIT for that file only.**

**2. Faster references / host-side work**
- `test_hopper_fp8_attention.py`: the seeded block-sparse BSR pattern + dense mask are cached per `(M, N, R, C)` (identical values, 24× reuse across heads/head_dim/dtype), and the eleven per-test 128 MB workspace allocations are replaced with one shared buffer. The sharing invariant (finish `plan()+run()` on one wrapper before planning another) is documented in the helper and verified at every call site: all tests complete the fp16 reference wrapper before the fp8 wrapper plans.
- `test_batch_prefill_kernels.py`: the per-request reference loops accumulate mismatch counts on-device and check once per test (previously up to 128 synchronizing `assert_close` calls per test), keeping request-sized comparison temporaries. The check asserts shapes up front (`isclose` broadcasts) and uses `equal_nan=False` to match `assert_close`'s default strictness; failures report the offending request indices.
- `test_fmha_v2_prefill.py`: `attention_ref_torch` / `chunked_attention_ref_torch` compute the same formula batched per KV-head group (one `[G, q_len, kv_len]` matmul per KV head instead of a Python loop over every query head), build position masks once per sequence, use in-place score ops to bound fp32 temporaries at long sequence lengths, and gather paged KV with one tensor index instead of a per-page `.item()` sync. This also cuts the reference's host RSS substantially.

**CI Time shrinkage**
* Current PR: `PR Test / JIT Unittest (H100) (pull_request)Successful in 237m`
* Adjacent PR #4509: `PR Test / JIT Unittest (H100) (pull_request)Successful in 292m`

Ten longest tests from logs
[Current PR:](https://github.com/flashinfer-ai/flashinfer/actions/runs/31755970069/job/94631720709?pr=4511)
```
Top 10 longest-running source files:
   1. tests/gdn/test_decode_delta_rule.py - duration 1h00m31s, peak RSS 2.8 GiB, peak GPU 41.1 GiB, samples 1290
   2. tests/attention/test_trtllm_gen_attention_decode_xqa.py - duration 18m42s, peak RSS 28.4 GiB, peak GPU 3.1 GiB, samples 374
   3. tests/attention/test_batch_attention.py - duration 18m27s, peak RSS 155.8 GiB, peak GPU 12.5 GiB, samples 284
   4. tests/attention/test_sliding_window.py - duration 16m55s, peak RSS 96.7 GiB, peak GPU 3.0 GiB, samples 318
   5. tests/attention/test_xqa.py - duration 15m01s, peak RSS 2.7 GiB, peak GPU 1002 MiB, samples 292
   6. tests/attention/test_xqa_batch_decode.py - duration 14m39s, peak RSS 2.7 GiB, peak GPU 4.7 GiB, samples 240
   7. tests/gemm/test_bmm_fp8.py - duration 14m08s, peak RSS 6.7 GiB, peak GPU 4.2 GiB, samples 336
   8. tests/attention/test_batch_prefill_kernels.py - duration 14m00s, peak RSS 44.2 GiB, peak GPU 18.9 GiB, samples 272
   9. tests/gemm/test_unified_gemm_fuzz.py - duration 11m47s, peak RSS 12.4 GiB, peak GPU 5.7 GiB, samples 289
  10. tests/attention/test_hopper.py - duration 11m18s, peak RSS 27.3 GiB, peak GPU 45.5 GiB, samples 248
```
[From PR #4509](https://github.com/flashinfer-ai/flashinfer/actions/runs/31745272862/job/94598305980?pr=4509)
```
Top 10 longest-running source files:
   1. tests/gdn/test_decode_delta_rule.py - duration 56m40s, peak RSS 2.7 GiB, peak GPU 41.1 GiB, samples 1281
   2. tests/attention/test_batch_prefill_kernels.py - duration 43m07s, peak RSS 49.6 GiB, peak GPU 19.0 GiB, samples 874
   3. tests/attention/test_fmha_v2_prefill.py - duration 20m13s, peak RSS 30.9 GiB, peak GPU 15.9 GiB, samples 218
   4. tests/attention/test_xqa.py - duration 19m38s, peak RSS 2.7 GiB, peak GPU 1002 MiB, samples 343
   5. tests/attention/test_sliding_window.py - duration 19m35s, peak RSS 182.4 GiB, peak GPU 3.0 GiB, samples 343
   6. tests/attention/test_trtllm_gen_attention_decode_xqa.py - duration 19m28s, peak RSS 28.0 GiB, peak GPU 3.1 GiB, samples 381
   7. tests/attention/test_batch_decode_kernels.py - duration 18m31s, peak RSS 67.0 GiB, peak GPU 63.7 GiB, samples 276
   8. tests/attention/test_hopper.py - duration 18m28s, peak RSS 26.6 GiB, peak GPU 49.5 GiB, samples 372
   9. tests/attention/test_hopper_fp8_attention.py - duration 18m12s, peak RSS 34.7 GiB, peak GPU 4.6 GiB, samples 397
  10. tests/attention/test_single_prefill.py - duration 15m38s, peak RSS 7.4 GiB, peak GPU 1.2 GiB, samples 264
```

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [ ] Tests have been added or updated as needed.
- [ ] All tests are passing (`unittest`, etc.).

## Tests

* Improved attention test performance by reducing repeated setup, synchronization, and memory allocation overhead.
* Enhanced validation for batch prefill and paged attention scenarios, including clearer reporting of mismatched requests and elements.
* Improved test reliability across supported hardware and configurations by handling unsupported cases more gracefully.
* Optimized reference implementations for paged and grouped attention while preserving expected results.

### CODE REVIEW

**Comment by bkryu** [comment]
@flashinfer-bot run

File: tests/attention/test_hopper_fp8_attention.py [hunk 1]
```diff
@@ -9,6 +10,38 @@
...
+    rng = np.random.default_rng(seed=0)
+    S = sp.sparse.random(MB, NB, density=0.25, random_state=rng).tocsr()
+    indptr = torch.from_numpy(S.indptr).cuda()
+    indices = torch.from_numpy(S.indices).cuda()
+    data_mask = torch.ones((S.nnz, R, C), dtype=torch.bool, device="cuda")
+    bsr = sp.sparse.bsr_matrix(
+        (np.ones((S.nnz, R, C), dtype=bool), S.indices, S.indptr), shape=(M, N)
+    )
+    dense_mask = torch.from_numpy(bsr.toarray()).to(device="cuda", dtype=torch.bool)
+    return indptr, indices, data_mask, dense_mask
```
[the remark this answers is not in the document]
Comment (bkryu) [reply tests/attention/test_hopper_fp8_attention.py:54]: Addressed in latest commit.

File: tests/attention/test_hopper_fp8_attention.py [hunk 2]
```diff
@@ -9,6 +10,38 @@
 from flashinfer.utils import is_sm90a_supported
 
 
+@functools.lru_cache(maxsize=2)
+def _workspace_buffer(size_mb: int = 128) -> torch.Tensor:
+    """Shared workspace buffer: allocating (and zeroing) 128 MB per test adds
+    up over thousands of parametrizations; the wrappers treat it as scratch."""
+    return torch.zeros(size_mb * 1024 * 1024, dtype=torch.uint8, device="cuda")
```
[the remark this answers is not in the document]
Comment (bkryu) [reply tests/attention/test_hopper_fp8_attention.py]: Addressed in latest commit.

File: tests/attention/test_fmha_v2_prefill.py [hunk 3]
```diff
@@ -132,68 +132,95 @@ def attention_ref_torch(
...
+            h_hi = h_lo + G
             if return_lse:
-                lse_seq[:, h] = torch.logsumexp(scores, dim=-1)
+                lse_seq[:, h_lo:h_hi] = torch.logsumexp(scores, dim=-1).permute(1, 0)
 
             attn = torch.softmax(scores, dim=-1)
-            o_seq[:, h, :] = torch.matmul(attn, v_h)
+            o_seq[:, h_lo:h_hi, :] = torch.matmul(attn, v_h.unsqueeze(0)).permute(
+                1, 0, 2
+            )
```
[the remark this answers is not in the document]
Comment (bkryu) [reply tests/attention/test_fmha_v2_prefill.py:223]: This amount of memory shouldn't be a big concern on H100

**Comment by bkryu** [comment]
@flashinfer-bot run

**Review by saltyminty — APPROVED** [review]

**Comment by saltyminty** [comment]
Will wait for CI to finish before merging.
